---
title: "Ex. 03-3: Integrating ChIP-seq results together"
output: 
    rmdformats::readthedown: 
        highlight: tango
        preserve_yaml: true
        df_print: tibble
        toc_depth: 4
        css: ../../../custom.css
        self_contained: false
    rmarkdown::html_document: 
        theme: simplex
        highlight: tango
        preserve_yaml: true
        code_folding: hide
        df_print: tibble
        toc: true
        toc_float: true
        self_contained: false
---

## 1. Connect to AWS

A general issue with bioinformatic analyses stems from the fact that nobody works in the same environment: 

- Are you working on Mac? Linux? Windows? 
- Do you have a lot of computational power? Perhaps a GPU card? 
- Are you connected to the Internet? With a fast connection? Are you working behind a proxy? 

To ensure that we are all working in the same environment, we will rely on AWS (Amazon Web Services) EC2 (Elastic Cloud 2) instances. 
EC2 instances are "virtual" computers to which you can connect remotely, from a local computer. Each one of you has been assigned a `user` ID (e.g. `user1`), 
along with a security "`key`" (e.g. `user1.pem`). Using these credentials, you can log in to the AWS instance dedicated for this course using command-line: 

> How to connect to AWS with a SSH client (`ssh ...` command line in Mac and Linux)? 

You will need your credentials (`user` ID), and you will also need to download the accompanying security `pem` file in your `Downloads` folder. 
Then, in your terminal, you can use the following snippet: 

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{sh eval = FALSE}
## Example
cd ~/Downloads/
chmod 600 "${USER}".pem
ssh -i "${USER}".pem "${USER}"@"${IP}"

## Actual Command Example
cd ~/Downloads/
chmod 600 user45.pem
ssh -i user45.pem user45@34.213.180.241
```
</p></details><br>

> How to connect to AWS without SSH client (generally for Windows users)? 

Things are a little bit more tedious for Windows users, but luckily not impossible!
There is a nice breakdown from another Physalia course on instructions for different operating systems and accessing AWS.
It is called [Connection to the Amazon EC2 service](https://gitlab.com/bfosso/metabarcoding_physalia/-/blob/master/unix_short_tutorial/how_to_connect.md). 
This will help with connecting to the AWS instance to run analyses.

## 2. Basic commands in AWS

Some basics commands in AWS are the following: 

- Check the your present directory

```{sh eval = FALSE}
pwd
```

- Check history

```{sh eval = FALSE}
history
```

- put history into a `history.txt` file

```{sh eval = FALSE}
history > history.txt
```

- make a new folder called data

```{sh eval = FALSE}
mkdir data
```

- Go to the new `data` directory

```{sh eval = FALSE}
cd data
```

- move `history.txt` file into `data` directory

```{sh eval = FALSE}
mv ../history.txt ./
```

- check manual page of `wget` command

```{sh eval = FALSE}
man wget
```

- redirect wget maunual page output into a file called `wget.txt`

```{sh eval = FALSE}
man wget > wget.txt
```

- return the lines that contain the "`output`" word in the `wget.txt` file

```{sh eval = FALSE}
cat wget.txt | grep output
grep -i output wget.txt
```

- Compress `wget.txt` file

```{sh eval = FALSE}
gzip wget.txt
```

- View Compressed file

```{sh eval = FALSE}
cat wget.txt.qz
zcat wget.txt.qz
zcat wget.txt.qz | less
```

- Download a file from Internet with `wget`

```{sh eval = FALSE}
wget ...rds
```

- List all files in a folder

```{sh eval = FALSE}
ls --color -Flh
```

## 3. RStudio

Most of single-cell RNA-seq analysis takes place either in `python` or in `R`. Here, we focus on how to leverage `R` to investigate scRNAseq data. 
`RStudio` is an `IDE` (Integrated Development Environment, in other words: a nice graphical interface to run `R`-related commands). 

We have installed `R` and `RStudio` on AWS. You can check which version of `R` is installed by opening an interactive `R` console: 

```{sh eval = FALSE}
R
```

Once you are within the `R` console, you can safely use `R` commands.

> Some useful commands in `R`:

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r eval = FALSE}
getwd() # equivalent of `pwd` in terminal
setwd("~/data/") # equivalent of `cd ~/data/` in terminal
list.files("~/data/") # equivalent of `ls` in terminal
download.file("...") # equivalent of `wget ...` in terminal
```
</p></details><br>

However, this can be quite cumbersome, and who likes an all-black screen with tiny letters and no handy features? 
So we can use `RStudio` (actually, `RStudio-server` since it is installed on an AWS remote server). 
Simply open a browser and copy-paste (the following address)[https://"${IP}":8787]: 

```{sh eval = FALSE}
https://"${IP}":8787
```

An `RStudio` log in page will appear; to log in, use your `user` ID for both ID and password. 

## 4. Installing packages 

So now, how do I install this or that brand new cutting-edge fancy "`package`"? 

`R` works with `packages`, available from different sources: 

- `CRAN`, the `R` developer team and official package provider: (CRAN)[https://cran.r-project.org/] (which can probably win the title of "the worst webpage ever designed in 1982"). 
- `Bioconductor`, another package provider, with a primary focus on genomic-related packages. 
-  Other unofficial sources, such as `GitHub`

> Install `mgcv`, `HCAData` and `revelio` packages

<details><summary style="color: #ff7f00; font-weight: bold">Show code</summary><p>
```{r eval = FALSE}
install.packages('mgcv')
BiocManager::install('HCAData')
devtools::install_github('danielschw188/revelio')
```
</p></details><br>

> Check package help pages

Package help pages are available at different places, depending on their source. 
That being said, there is a place I like to go to easily find information related to most packages: 

(https://rdrr.io/)[https://rdrr.io/]

Check out `Revelio` package help pages. 

> Can you speculate on how the transcription factors combine on the chromatin? 

> What was your initial hypothesis regarding the different types of peak sets (Ex. 03-1)? Does this match your observations here? 